components/sup/src/http_gateway.rs

```diff
@@ -142,6 +149,58 @@ impl AppState {
     }
 }
 
+// Our authentication middleware
+struct Authentication;
+
+impl Middleware<AppState> for Authentication {
+    fn start(&self, req: &HttpRequest<AppState>) -> actix_web::Result<Started> {
+        let current_token = &req
+            .state()
+            .gateway_state
+            .read()
+            .expect("GatewayState lock is poisoned")
+            .auth_token;
+
+        let current_token = match current_token.as_ref() {
+            Some(t) => t,
+            // If there's no auth token in the state, just return. Everything will continue to function
+            // unauthenticated.
+            None => {
+                debug!("No auth token present. HTTP gateway starting in unauthenticated mode.");
+                return Ok(Started::Done);
+            }
+        };
+
+        // From this point forward, we know that we have an auth token in the state. Therefore,
+        // anything short of a fully formed Authorization header containing a Bearer token that
+        // matches the value we have in our state, results in an Unauthorized response.
+
+        let hdr = match req.headers().get(http::header::AUTHORIZATION) {
+            Some(hdr) => {
+                match hdr.to_str() {
+                    Ok(h) => h,
+                    Err(e) => {
+                        debug!("Authorization headers can only contain printable ASCII characters. {:?}.", e);
+                        return Ok(Started::Response(HttpResponse::Unauthorized().finish()));
+                    }
+                }
+            }
+            None => return Ok(Started::Response(HttpResponse::Unauthorized().finish())),
+        };
```

Comment:
Thanks for making the change anyway. I think the more FP-oriented approach *is* indeed a bit harder to read until one gets used to it, but I think it pays dividends in the long run since it makes certain classes of errors get caught by the compiler or simply impossible.